backup, jobs: reliably emit context cancelled on drain, treat as retryable

dt opened the pull request:
Closes #156337.

dt committed:
distsql: emit cancellation error in MoveToDraining

Previously if the context used to run a processor was canceled, most of
the time the emitted error would be that context cancellation, however
this depended on the exact implementation of each processor and the code
it called reliably returning the context cancelled error exactly in this
case. In at least some places, this is not reliably done.

This changes DistSQL's base processor to ensure that if the context it
is using to run a processor is canceled, the error emitted in MoveToDraining
is always the context cancellation error, by overriding the error passed
with the context error in MoveToDraining if the former is set.

This ensures callers inspecting errors to determine if they are due to
cancellation can always rely a cancelled flow's emitted error being
a cancellation error.

Fixes #156337.

Release note: none.
Epic: none.

dt committed:
jobs: mark context.Canceled as retryable

Release note: none.
Epic: none.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @yuzefovich reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157145#-:-Odj_9IeB-rZQidB51uL:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jeffswenson and @ZhouXing19)

___
*[`pkg/sql/execinfra/processorsbase.go` line 528 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157145#-OdjZg4H3vcArh9DmOZ8:-OdjZg4H3vcArh9DmOZ9:b1dm32o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6993aeacdcc7ad15e958cdae133b8ec8fbef242b/pkg/sql/execinfra/processorsbase.go#L528)):*
> ```Go
> // at init() time), then we move straight to the StateTrailingMeta.
> //
> // An error can be optionally passed. It will be the first piece of metadata
> ```

nit: let's extend this comment a bit.
___
*[`pkg/sql/execinfra/processorsbase.go` line 557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157145#-OdjZtNvArrS3doHpXd4:-OdjZtNvArrS3doHpXd5:b-i0o3mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6993aeacdcc7ad15e958cdae133b8ec8fbef242b/pkg/sql/execinfra/processorsbase.go#L557)):*
> ```Go
> 		// reflect that, making cancellation detectable by callers.
> 		if pb.ctx.Err() != nil {
> 			if !errors.Is(err, pb.ctx.Err()) {
> ```

nit: use `pb.Ctx()` here and below.


<!-- Sent from Reviewable.io -->

dt committed:
changefeedcccl: move to draining before cancelling proc

Once a proc's context has been cancelled, it may cease processing emitted rows or even
emitted drain transition information, so if the processor implementation wishes to
reliably emit information about why it is moving to draining, it should do so before
it cancels its underlying base processor.

Release note: none.
Epic: none.

dt commented:
> nit: let's extend this comment a bit.

Done.

> nit: use pb.Ctx() here and below.

Done.

yuzefovich reviewed (approved):
> @yuzefovich reviewed 1 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jeffswenson, @log-head, @msbutler, and @ZhouXing19)



<!-- Sent from Reviewable.io -->

dt commented:
TFTR!

bors r+
